### DISCUSSION TITLE
Migrating to 0.7.0 with Axum and Extractors

### DISCUSSION THREAD

**Discussion by lukehsiao** [leptos-rs/leptos#3366] [Q&A] [answered]
I have an app I'm trying to migrate to leptos 0.7. Full code is available here: https://github.com/lukehsiao/find-the-password

```
    let conf = get_configuration(None).unwrap();
    let leptos_options = conf.leptos_options;
    let addr = leptos_options.site_addr;
    let routes = generate_route_list(App);
    let app_state = Internal {
        leptos_options,
        usermap: Arc::new(Users::new()),
        leaderboard: Arc::new(Mutex::new(vec![])),
    };

    // build our application with a route
    let app = Router::new()
        .route("/u/:user/check/:password", routing::get(check_password))
        .route("/u/:user/passwords.txt", routing::get(get_passwords))
        .leptos_routes_with_context(
            &app_state,
            routes,
            {
                let app_state = app_state.clone();
                move || provide_context(app_state.clone())
            },
            App,
        )
        .fallback(file_and_error_handler)
        .with_state(app_state)
        .layer(CompressionLayer::new());
```

I'm running into confusion of how to handle the new `shell`. Specifically, after making the changes I think I need to make so it can compile (reproducible here: https://github.com/lukehsiao/find-the-password/pull/227)

I see the error below when connecting via my web browser.

```
2024-12-15T07:42:33.303494Z TRACE axum::serve: connection 127.0.0.1:35176 accepted
panicked at /home/lukehsiao/.cargo/registry/src/index.crates.io-6f17d22bba15001f/leptos_meta-0.7.0/src/lib.rs:247:18:
you are using leptos_meta without a </head> tag
```

which I assume is because I'm not using the new `shell`.

I'm confused how to do so with `leptos_routes_with_context`, and I don't see any existing examples which use this leptos routes function. How can I provide_context and get my Axum extractor state, but also configure the shell?

Can anyone give me some pointers?

**Comment by lukehsiao** [comment, marked as answer]
Ah, turns out I was just looking in the wrong directory.

I found all the examples I needed in this: https://github.com/leptos-rs/leptos/tree/68fdc8d9c496d9ba85c2ea1646d41cb1be4638ed/projects/session_auth_axum

Thanks!

**Reply by gbj** [reply]
Great!
